File: go/e2e/sorted_set_redis_test.go
//go:build e2e

package e2e

import (
	"context"
	"testing"
)

func TestRedisSortedSet_HeartbeatLifecycle(t *testing.T) {
	flushRedis(t)
	store := createTestStore(t)
	ctx := context.Background()

	setName := "cluster:heartbeats"

	// 1. Add heartbeats
	if err := store.SortedSet().Add(ctx, setName, "node-1", 1000.0); err != nil {
		t.Fatalf("Add failed: %v", err)
	}
	if err := store.SortedSet().Add(ctx, setName, "node-2", 1050.0); err != nil {
		t.Fatalf("Add failed: %v", err)
	}
	if err := store.SortedSet().Add(ctx, setName, "node-3", 1100.0); err != nil {
		t.Fatalf("Add failed: %v", err)
	}

	// 2. Count
	count, err := store.SortedSet().Count(ctx, setName)
	if err != nil || count != 3 {
		t.Fatalf("Count failed: count=%d, err=%v", count, err)
	}

	// 3. Score
	score, err := store.SortedSet().Score(ctx, setName, "node-2")
	if err != nil || score == nil || *score != 1050.0 {
		t.Fatalf("Score failed: score=%v, err=%v", score, err)
	}

	scoreMissing, err := store.SortedSet().Score(ctx, setName, "node-unknown")
	if err != nil || scoreMissing != nil {
		t.Fatalf("Score for missing failed: score=%v, err=%v", scoreMissing, err)
	}

	// 4. RangeByScore
	active, err := store.SortedSet().RangeByScore(ctx, setName, 1040.0, 1110.0, nil)
	if err != nil || len(active) != 2 {
		t.Fatalf("RangeByScore failed: len=%d, err=%v", len(active), err)
	}
	if active[0].Member != "node-2" || active[1].Member != "node-3" {
		t.Fatalf("RangeByScore unexpected members: %+v", active)
	}

	// 5. RemoveRange (nodes older than 1020)
	removed, err := store.SortedSet().RemoveRange(ctx, setName, 0.0, 1020.0)
	if err != nil || removed != 1 {
		t.Fatalf("RemoveRange failed: removed=%d, err=%v", removed, err)
	}

	remCount, _ := store.SortedSet().Count(ctx, setName)
	if remCount != 2 {
		t.Fatalf("expected 2 remaining, got %d", remCount)
	}

	// 6. Remove
	if err := store.SortedSet().Remove(ctx, setName, "node-3"); err != nil {
		t.Fatalf("Remove failed: %v", err)
	}

	finalCount, _ := store.SortedSet().Count(ctx, setName)
	if finalCount != 1 {
		t.Fatalf("expected 1 remaining, got %d", finalCount)
	}
}

File: rust/tests/sorted_set_redis_test.rs
use super::common::*;

#[tokio::test]
async fn test_redis_sorted_set_heartbeat_lifecycle() {
    flush_redis().await.expect("failed to flush redis");
    let store = create_test_store().await.expect("failed to create store");

    let set_name = "cluster:heartbeats";

    // 1. Add heartbeats with unix timestamps
    store
        .sorted_set
        .add(set_name, "node-1", 1000.0)
        .await
        .unwrap();
    store
        .sorted_set
        .add(set_name, "node-2", 1050.0)
        .await
        .unwrap();
    store
        .sorted_set
        .add(set_name, "node-3", 1100.0)
        .await
        .unwrap();

    // 2. Count
    let total = store.sorted_set.count(set_name).await.unwrap();
    assert_eq!(total, 3);

    // 3. Score
    let score = store.sorted_set.score(set_name, "node-2").await.unwrap();
    assert_eq!(score, Some(1050.0));

    let score_missing = store
        .sorted_set
        .score(set_name, "node-unknown")
        .await
        .unwrap();
    assert_eq!(score_missing, None);

    // 4. Range by score (find active nodes between 1040 and 1110)
    let active = store
        .sorted_set
        .range_by_score(set_name, 1040.0, 1110.0, None)
        .await
        .unwrap();
    assert_eq!(active.len(), 2);
    assert_eq!(active[0].0, "node-2");
    assert_eq!(active[1].0, "node-3");

    // 5. Remove expired nodes (older than 1020)
    let removed = store
        .sorted_set
        .remove_range(set_name, 0.0, 1020.0)
        .await
        .unwrap();
    assert_eq!(removed, 1);

    let remaining = store.sorted_set.count(set_name).await.unwrap();
    assert_eq!(remaining, 2);

    // 6. Explicit remove
    store.sorted_set.remove(set_name, "node-3").await.unwrap();
    assert_eq!(store.sorted_set.count(set_name).await.unwrap(), 1);
}
